File: artifacts/tulipindicators/translate_new_ablation/src/linregslope.rs
use crate::*;
use rand::Rng;

pub fn ti_linregslope_start(options: Option<&[f64]>) -> i32 {
    match options {
        Some(opts) if !opts.is_empty() => opts[0] as i32 - 1,
        _ => -1, // Return -1 for None or empty slice (equivalent to NULL/invalid pointer)
    }
}

pub fn ti_linregslope(size: i32, inputs: &[&[f64]], options: &[f64], outputs: &mut [&mut [f64]]) -> i32 {
    let input = inputs[0];
    let period = options[0] as i32;
    let output = &mut outputs[0];
    
    if period < 1 {
        return 1;
    }
    
    let start = ti_linregslope_start(Some(options));
    if size <= start {
        return 0;
    }
    
    let mut x = 0.0;
    let mut x2 = 0.0;
    let mut y = 0.0;
    let mut xy = 0.0;
    
    for i in 0..(period - 1) {
        let weight = (i + 1) as f64;
        x += weight;
        x2 += weight * weight;
        xy += input[i as usize] * weight;
        y += input[i as usize];
    }
    
    let period_f64 = period as f64;
    x += period_f64;
    x2 += period_f64 * period_f64;
    let bd = 1.0 / ((period_f64 * x2) - (x * x));
    
    let mut output_idx = 0;
    for i in (period - 1)..size {
        let i_usize = i as usize;
        xy += input[i_usize] * period_f64;
        y += input[i_usize];
        
        let b = ((period_f64 * xy) - (x * y)) * bd;
        output[output_idx] = b;
        output_idx += 1;
        
        xy -= y;
        y -= input[(i - period + 1) as usize];
    }
    
    assert!((output_idx as i32) == (size - start));
    0
}

File: artifacts/tulipindicators/decompose/src/linregslope.c
int ti_linregslope_start(const double *options)
{
  return ((int) options[0]) - 1;
}


/*** DEPENDENCIES:
***/


int ti_linregslope(int size, const double * const *inputs, const double *options, double * const *outputs)
{
  const double *input = inputs[0];
  const int period = (int) options[0];
  double *output = outputs[0];
  unsigned int output_idx = 0;
  if (period < 1)
  {
    return 1;
  }
  if (size <= ti_linregslope_start(options))
  {
    return 0;
  }
  do
  {
    double x = 0;
    double x2 = 0;
    double y = 0;
    double xy = 0;
    do
    {
    }
    while (0);
    int i;
    for (i = 0; i < (period - 1); i += 1)
    {
      x += i + 1;
      x2 += (i + 1) * (i + 1);
      xy += input[i] * (i + 1);
      y += input[i];
    }

    x += period;
    x2 += period * period;
    const double bd = 1.0 / ((period * x2) - (x * x));
    for (i = period - 1; i < size; i += 1)
    {
      xy += input[i] * period;
      y += input[i];
      const double b = ((period * xy) - (x * y)) * bd;
      do
      {
        output[output_idx] = b;
        output_idx += 1;
      }
      while (0);
      xy -= y;
      y -= input[(i - period) + 1];
    }

  }
  while (0);
  assert(((&output[output_idx]) - outputs[0]) == (size - ti_linregslope_start(options)));
  return 0;
}


/*** DEPENDENCIES:
int ti_linregslope_start(const double *options)
{
  return ((int) options[0]) - 1;
}


----------------------------
***/


